library/bignum.c

```diff
@@ -1877,19 +1877,31 @@ int mbedtls_mpi_mod_int( mbedtls_mpi_uint *r, const mbedtls_mpi *A, mbedtls_mpi_
 
 /*
  * Fast Montgomery initialization (thanks to Tom St Denis)
+ *
+ * This computes the negative modular inverse of m in mbedtls_mpi_uint.
  */
-static void mpi_montg_init( mbedtls_mpi_uint *mm, const mbedtls_mpi *N )
-{
-    mbedtls_mpi_uint x, m0 = N->p[0];
-    unsigned int i;
-
-    x  = m0;
-    x += ( ( m0 + 2 ) & 4 ) << 1;
-
-    for( i = biL; i >= 8; i /= 2 )
-        x *= ( 2 - ( m0 * x ) );
-
-    *mm = ~x + 1;
+static void mpi_montg_init( mbedtls_mpi_uint *m_inv,
+                            mbedtls_mpi_uint const *m )
+{
+    /* The basic principle is the following:
+     * If R=2^k and we already know that m_inv is the negative
+     * modular inverse of m w.r.t. R -- that is,
+     *   m_inv*m + 1 == 0 mod R,
+     * then
+     *  (m_inv*m + 1)^2 == 0 mod R^2.
+     * Expanding yields that
+     *  2 m_inv + m_inv^2 * m = m_inv * (2 + m_inv*m)
+     * is the modular inverse of -m w.r.t. R^2 = 2^{2k}
+     *
+     * Below, we iterate this procedure, starting at the bitlevel (k=1)
+     * where m_inv = m = 1 (m must be odd). We compute in mbedtls_mpi_uint
+     * throughout, which is higher precision than necessary for most of
+     * the iterations, but yields the simplest and fastest code.
```

Comment:
(NB: Alluding to speed is perhaps misleading here -- the cost of this routine is irrelevant since it's only executed once within an RSA operation)